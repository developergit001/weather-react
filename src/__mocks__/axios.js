/*
WARNING: If you use import mockAxios from "../__mocks__/axios";
then you must be careful, because ALL the test will be using the face axios with THIS default data
(defaultdata = {}) and then of course, not all your test component that use real axios would be receive
this response that the correct one of the real axios, so probably its gonna fail some test because some
components must expect another response, not this one, so... for global mock axios it would be a problem maybe.
*/

//Fake data. if you have two axios request and expect distincts data response, then this solution its a problem.
//You must fake all the testing using a diferent mock in each testing.
//const defaultdata = {"cod":0,"entity":{"status":"success","country":"Argentina","countryCode":"AR","region":"B","regionName":"Buenos Aires","city":"Tablada","zip":"1766","lat":-34.7065,"lon":-58.5272,"timezone":"America/Argentina/Buenos_Aires","isp":"Telecentro S.A.","org":"Telecentro S.A. - Clientes Residenciales","as":"AS27747 Telecentro S.A.","query":"181.46.66.72"}};
const defaultdata = {};

export default {
    get: jest.fn(() => Promise.resolve({ data: defaultdata }))
};