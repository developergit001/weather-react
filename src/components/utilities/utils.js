
//Return the param x with trim regex applied.
function myTrim(x) {
    return x.replace(/^\s+|\s+$/gm,'');
}
function getCentrigrados(gradosKelvin){
    //return (gradosF !== '')?(((gradosF - 32) * 5)/9):gradosF;
    return (gradosKelvin !== '')?(Number(gradosKelvin) - 273.15).toFixed(0):gradosKelvin;    
}
function getBackUrl(){
    return "http://localhost:8080";
}
module.exports = { 
    getCentrigrados,getBackUrl,myTrim
}