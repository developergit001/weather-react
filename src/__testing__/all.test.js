import React from 'react';
import renderer from 'react-test-renderer';
import {render, cleanup, waitForElement} from '@testing-library/react';
import Weather from "../components/weather";
import { getCentrigrados, getWeather, getWeatherEndPoint } from '../components/utilities/utils';
import Main from '../components/main';

/*
If you want that all your tests (each one of them) use the fake mock axios 
then USE:
Choice A)
import mockAxios from "../__mocks__/axios"; //for global response with exact the same json inside here!

otherwise USE this 2 lines:

CHOICE B)
import axios from "axios";
jest.mock('axios');
*/

//CHOICE B)
import axios from "axios";
jest.mock('axios');


afterEach(cleanup);

/*
PRUEBAS: Este archivo tiene realizada todas las posibles pruebas a modo de ejemplo para la demo
  Tipos de pruebas: 
  Pruebas de snapshot para evaluar y verificar cambios de componentes a nivel UI.
  Pruebas unitarias, para evaluar resultados en funciones
  Pruebas usando mocks/fakes functions para emular peticiones ajax
  Pruebas de contenido a nivel arbol de dom, para evaluar si ciertos componentes contiene los datos esperados.
  Al realizar los dos desarrollos frontend en react y nodejs, el testing se realizó solo en este proyecto,
  ya que las pruebas de backend serian similares y en menor cantidad ya que sólo se hacen request en el back.
*/

it('Prueba de mock fake testing simulando ajax request', async () => {

  //Dato de respuesta emulado
  const resdata = {data:{"cod":0,"mockaxiostest":true,"entity":{"coord":{"lon":12.48,"lat":41.89},"weather":[{"id":800,"main":"Clear","description":"cielo claro","icon":"01d"}],"base":"stations","main":{"temp":295.62,"feels_like":296.88,"temp_min":294.26,"temp_max":298.15,"pressure":1011,"humidity":83},"visibility":10000,"wind":{"speed":3.1,"deg":130},"clouds":{"all":0},"dt":1593748201,"sys":{"type":1,"id":6792,"country":"IT","sunrise":1593747586,"sunset":1593802118},"timezone":7200,"id":3169070,"name":"Rome","cod":200}}};
  // setup
  /*
  //CHOICE A
  mockAxios.get.mockImplementationOnce(() =>
    Promise.resolve(resdata)
  );
  */

  /*
  //First, by default the new mock axios return undefined, you must put a default data return.

  axios.get('https://www.google.com'); //undefined!!!
  const datos = await getWeather(cityId); //undefined too!!!
  */

  //Doing this we put a default value for mock
  axios.get.mockResolvedValue(
    resdata //The data that i must test
  );


  const cityId = "3169070"; //pasamos como city a Roma.
  let endpointurl = getWeatherEndPoint() + "/" + cityId;

  const datos = await getWeather(cityId);
  
  //console.log("mock result",datos); //Call the mock axios yeah!!!
  
  expect(datos).toEqual(resdata);
  
  //IF we use __mock__/axios.js in every test, replace the real axios for the fake one
  //__mock__/axios.js has a default data too, must be the correct want for every test, be carefull.
  //CHOICE A)
  //expect(mockAxios.get).toHaveBeenCalledTimes(1);
  //expect(mockAxios.get).toHaveBeenCalledWith(endpointurl /*,{params: {cityId:cityId}}*/);
  
  //CHOICE B) only for this test.
  expect(axios.get).toHaveBeenCalledTimes(1);
  expect(axios.get).toHaveBeenCalledWith(endpointurl /*,{params: {cityId:cityId}}*/);

});

it('Unit Test fx getCentrigrados ', () => {
    //Conversion Kelvin 2 grados centigrados. (sin decimales)
    expect(getCentrigrados(283.15)).toBe("10");
});

//Datos basados de openweatherapi
const tweather = {"coord":{"lon":-58.53,"lat":-34.71},"weather":[{"id":803,"main":"Clouds","description":"nubes rotas","icon":"04d"}],"base":"stations","main":{"temp":284.13,"feels_like":280.54,"temp_min":283.15,"temp_max":284.82,"pressure":1020,"humidity":76},"visibility":10000,"wind":{"speed":4.1,"deg":290},"clouds":{"all":75},"dt":1593711132,"sys":{"type":1,"id":8237,"country":"AR","sunrise":1593687713,"sunset":1593723259},"timezone":-10800,"id":3436426,"name":"Aldo Bonzi","cod":200};

it("Componente Weather: imagen de icono de clima, bien armada",()=>{

  const imgurl = 'http://openweathermap.org/img/wn/' + tweather.weather[0].icon + '@2x.png';
  const {getByTestId} = render(<Weather data={tweather} isLoading={false} ahora="(Clima actual)" ></Weather>)
  expect(getByTestId("icon-weather-img").getAttribute("src")).toBe(imgurl);

});

it("Test Weather component SNAPSHOT Test",()=>{

    //PRIMER SNAPSHOT
    //Componente vacio, esta cargando... sin datos, ERGO muestra una imagen de loading...
    const component1 = renderer.create(
      <Weather data={{}} isLoading={true} ahora="" ></Weather>,
    );
    let tree = component1.toJSON();
    expect(tree).toMatchSnapshot(); //Genera el snapshot / HTML
  
    //SIGUIENTE SNAPSHOT
    //Componente con datos, en esta instancia el <img loading  > NO se muestra...  
    const component2 = renderer.create(
      <Weather data={tweather} isLoading={false} ahora="(Clima actual)" ></Weather>,
    );
  
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot(); //Generando segundo snapshot.
  
});

it('Prueba de contenido Existe un mensaje de bienvenida en la app', async () => {
    const { getByText } = render(<Main />);
    //const linkElement = getByText(/Bienvenido/i);
    const linkElement = await waitForElement(() => getByText(/Bienvenido/i));
    expect(linkElement).toBeInTheDocument();
});

/*
SIN USO
test('Test Weather component SNAPSHOT Test', () => {
  //Cosas que podriamos hacer con el tree DOM, por ejemplo llamar a una funcion, en nuestro caso este componente no hace nada.
  //Pero podria darse el caso que un evento onMouseEnter cambie el class de un HTML y queremos ver si se impacta.
  
  // manually trigger the callback
  //tree.props.onMouseEnter();
  // re-rendering
  //tree = component.toJSON();
  //expect(tree).toMatchSnapshot();

  // manually trigger the callback
  //tree.props.onMouseLeave();
  // re-rendering
  //tree = component.toJSON();
  //expect(tree).toMatchSnapshot();

});
*/