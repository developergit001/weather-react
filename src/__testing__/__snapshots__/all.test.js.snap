// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Weather component SNAPSHOT Test 1`] = `
<div
  className="weather"
>
  <img
    alt="Cargando..."
    src="Rolling-1s-200px.svg"
  />
  <br />
  Cargando...
</div>
`;

exports[`Test Weather component SNAPSHOT Test 2`] = `
<div
  className="weather"
>
  <div
    className="weather-header"
  />
  <div
    className="weather-body"
  >
    <div
      className="weather-body__img"
    >
      <img
        alt="weather icon"
        data-testid="icon-weather-img"
        src="http://openweathermap.org/img/wn/04d@2x.png"
      />
    </div>
    <div
      className="weather-body__degree"
    >
      11
      Â°
    </div>
  </div>
  <div
    className="weather-footer"
  >
    <div
      className="weather-footer__info"
    >
      nubes rotas
    </div>
    <div
      className="weather-footer__date"
    >
      (Clima actual)
    </div>
  </div>
</div>
`;
